sherpa/astro/tests/test_astro_data.py

```diff
@@ -2953,8 +2974,12 @@ def test_pha_related_field_can_not_be_a_set_sequence(field):
     data_class, args = PHA_ARGS
     data = data_class(*args)
 
-    # this does not fail
-    setattr(data, field, [{"abc", False, 23.4}] * 3)
+    with pytest.raises(DataErr) as err:
+        # XFAIL does not raise an error except for grouping/quality
+        setattr(data, field, [{"abc", False, 23.4}] * 3)
+
+    # once non-quality/grouping fields fail the error check will need to be changed
+    assert str(err.value) == "Array must be a sequence of integers or None"
```

Comment:
I don't think we want syserror to be forced to be an integer, which is what this error message suggests. Aha. The syserror (in fact, most fields being tested here) actually do not trigger this warning - i.e they are marked as xfail. Aha, this gets resolved in a later commit so the error message is then "Array must be a sequence or None".

We could support setting the systematic error to a scalar (e.g. 0.05 to mean 5%, which then gets applied to the existing data) - and in fact I think the UI does support this (would need to check, but I seem to remember it doing so). There are issues to resolve: i.e. if I've set it to 0.05 and then I've changed the y/dependent axis, should it know to automatically apply this to the new data or require the call to be re-run (arguments are possible for all sides).

Reply:
The latter (set it to the scalar 0.05) is what I had in mind. I need to read up on the docs, but intuitively, I had expected it to be an absolute number, not a fraction (i.e. 0.05 is 0.05, which can be a lot as in 0.01 +/- 0.05, or a little, as in 1234.566 +\- 0.05)
I don't think it's required to reset is when the dependent axis changes. To me, that's a property of the dataset. Since the independent axis is not harder to change, I would say that "I'm have a totally different measurement and I'm just re-using the same object" is not the prime use case we should be worried about.

Reply:
That said, if it's much easier to reset it  when the dependent axis is changed, I can work with that, too. As @DougBurke and agument can be made either way, we just have to write down in the docs what choice we made.

Reply:
So, we have different behavior in the UI layer to the data object. In the UI layer you can call `set_staterror` or `set_syserror` to a scalar and it replicates the value for each bin [*]. If you manually change the `staterror` or `syserror` field of the data object then in CIAO 4.14 it just sets the value to a scalar, which technically may work (what with broadcasting and all) but this PR is all about trying to make the code a bit easier to write.

We could move the logic that's currently in `ui.set_staterror`/`ui.set_staterror` into the `staterror`/`syserror` properties (it would fit into my desire to reduce the logic in the ui code).

Here's an example from a CIAO 4.14 sherpa session:

```
sherpa-4.14.0> load_arrays(1, [1, 2, 3], [0, 1, -1])

sherpa-4.14.0> get_staterror()
array([1.8660254 , 2.32287566, 1.8660254 ])

sherpa-4.14.0> get_syserror()
DataErr: data set '1' does not specify systematic errors

sherpa-4.14.0> set_syserror(0.5)

sherpa-4.14.0> get_syserror()
array([0.5, 0.5, 0.5])

sherpa-4.14.0> set_staterror(1.2)

sherpa-4.14.0> get_staterror()
array([1.2, 1.2, 1.2])

sherpa-4.14.0> data = get_data()

sherpa-4.14.0> data.staterror = 2

sherpa-4.14.0> data.syserror = 3

sherpa-4.14.0> print(data)
name      = 
x         = Int64[3]
y         = Int64[3]
staterror = 2
syserror  = 3
```

[*] note to self: what does it do for grouped PA data

Reply:
If you look in `sherpa.ui/utils.py` (not going to link to it as it'll just get rebased away soon...) one of the things I did was move some code related to setting dependent axis/errors/filters into separate routines that could be called from the `Session` class. You can see below that we have logic in the ui layer to allow some of these fields to be set to a scalar, but they are not in the `Data` methods. I'd prefer to resolve this in a separate, targeted, PR rather than this one.

```
def set_dep(data, val):
    """Set the dependent axis.

    Parameters
    ----------
    data : sherpa.data.Data instance
    val : scalar or array_like
        If a scalar then use the same value for all elements.

    """

    if numpy.iterable(val):
        dep = numpy.asarray(val, SherpaFloat)
    else:
        val = SherpaFloat(val)
        dep = numpy.array([val] * data.size)

    data.dep = dep


def set_error(data, field, val, fractional=False):
    """Set the error field.

    Parameters
    ----------
    data : sherpa.data.Data instance
    field : {"staterror", "syserror"}
    val : None, a scalar, or array_like
        The statistical error.
    fractional : bool, optional
        If `val` is a scalar then set errors to
        `val * data.y`.

    Notes
    -----
    The `val` field, when an array, must match the data.size
    field, even for grouped PHA data.

    """

    if val is None:
        err = None

    elif numpy.iterable(val):
        err = numpy.asarray(val, SherpaFloat)

    else:
        val = SherpaFloat(val)
        if sherpa.utils.bool_cast(fractional):
            err = val * data.get_dep()
        else:
            err = numpy.array([val] * data.size)

    setattr(data, field, err)


def set_filter(data, val, ignore=False):
    """Set the filter field.

    Parameters
    ----------
    data : sherpa.data.Data instance
    val : array_like
        The filter array (bool-like).
    ignore : bool, optional
        If set the filter should be inverted.

    Notes
    -----
    The `val` field should match the mask array, that is the
    "effective" size of the data. This is only different from
    data.size when the data is grouped (and at least one group
    contains multiple channels).

    """

    val = numpy.asarray(val, dtype=numpy.bool_)
    nval = len(val)

    # Note that we do not use data.size as the check, as that fails
    # for grouped PHA data. Is there some way to have an "effective
    # size" value?
    #
    nexp = len(data.get_y(False))
    if numpy.iterable(data.mask):
        if nexp != nval:
            raise sherpa.utils.err.DataErr('mismatchn', 'data', 'filter',
                                           nexp, nval)

        if not ignore:
            data.mask |= val
        else:
            data.mask &= ~val

        return

    # The mask attribute checks the length, so we do not have to here.
    #
    if not ignore:
        data.mask = val
    else:
        data.mask = ~val
```